File: services/menos/menos/services/llm_providers.py
"""Cloud LLM provider implementations for OpenAI, Anthropic, and OpenRouter."""

import asyncio
import logging
from typing import Any

import httpx

from menos.services.llm import LLMProvider

logger = logging.getLogger(__name__)


class OpenAIProvider:
    """LLM provider implementation using OpenAI Chat Completions API.

    Uses the standard chat completions endpoint with exponential backoff retry logic.
    """

    def __init__(self, api_key: str, model: str = "gpt-4o-mini"):
        """Initialize OpenAI provider.

        Args:
            api_key: OpenAI API key
            model: Model identifier (e.g., gpt-4o-mini, gpt-4-turbo)
        """
        self.api_key = api_key
        self.model = model
        self.client: httpx.AsyncClient | None = None
        self.base_url = "https://api.openai.com/v1"

    async def _get_client(self) -> httpx.AsyncClient:
        """Lazy initialize HTTP client with auth headers."""
        if self.client is None:
            self.client = httpx.AsyncClient(
                base_url=self.base_url,
                headers={
                    "Authorization": f"Bearer {self.api_key}",
                    "Content-Type": "application/json",
                },
            )
        return self.client

    async def generate(
        self,
        prompt: str,
        *,
        system_prompt: str | None = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
        timeout: float = 60.0,
    ) -> str:
        """Generate text using OpenAI Chat Completions API.

        Args:
            prompt: The user message/prompt
            system_prompt: Optional system message to guide generation
            max_tokens: Maximum tokens to generate
            temperature: Sampling temperature (0.0-2.0)
            timeout: Request timeout in seconds

        Returns:
            Generated text response

        Raises:
            RuntimeError: If generation fails after all retries
        """
        max_retries = 3
        base_delay = 1.0

        for attempt in range(max_retries):
            try:
                client = await self._get_client()

                messages: list[dict[str, str]] = []
                if system_prompt is not None:
                    messages.append({"role": "system", "content": system_prompt})
                messages.append({"role": "user", "content": prompt})

                payload: dict[str, Any] = {
                    "model": self.model,
                    "messages": messages,
                    "max_tokens": max_tokens,
                    "temperature": temperature,
                }

                response = await client.post(
                    "/chat/completions",
                    json=payload,
                    timeout=timeout,
                )
                response.raise_for_status()
                data = response.json()

                return data["choices"][0]["message"]["content"]

            except httpx.HTTPError as e:
                if attempt == max_retries - 1:
                    raise RuntimeError(
                        f"OpenAI generation failed after {max_retries} retries: {e}"
                    ) from e

                delay = base_delay * (2**attempt)
                await asyncio.sleep(delay)

        return ""

    async def close(self) -> None:
        """Close the HTTP client and cleanup resources."""
        if self.client is not None:
            await self.client.aclose()
            self.client = None


class AnthropicProvider:
    """LLM provider implementation using Anthropic Messages API.

    Uses the Messages API with required anthropic-version header.
    """

    def __init__(self, api_key: str, model: str = "claude-3-5-haiku-20241022"):
        """Initialize Anthropic provider.

        Args:
            api_key: Anthropic API key
            model: Model identifier (e.g., claude-3-5-haiku-20241022, claude-3-5-sonnet-20241022)
        """
        self.api_key = api_key
        self.model = model
        self.client: httpx.AsyncClient | None = None
        self.base_url = "https://api.anthropic.com/v1"

    async def _get_client(self) -> httpx.AsyncClient:
        """Lazy initialize HTTP client with auth and version headers."""
        if self.client is None:
            self.client = httpx.AsyncClient(
                base_url=self.base_url,
                headers={
                    "x-api-key": self.api_key,
                    "anthropic-version": "2023-06-01",
                    "Content-Type": "application/json",
                },
            )
        return self.client

    async def generate(
        self,
        prompt: str,
        *,
        system_prompt: str | None = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
        timeout: float = 60.0,
    ) -> str:
        """Generate text using Anthropic Messages API.

        Args:
            prompt: The user message/prompt
            system_prompt: Optional system prompt to guide generation
            max_tokens: Maximum tokens to generate
            temperature: Sampling temperature (0.0-2.0)
            timeout: Request timeout in seconds

        Returns:
            Generated text response

        Raises:
            RuntimeError: If generation fails after all retries
        """
        max_retries = 3
        base_delay = 1.0

        for attempt in range(max_retries):
            try:
                client = await self._get_client()

                payload: dict[str, Any] = {
                    "model": self.model,
                    "messages": [{"role": "user", "content": prompt}],
                    "max_tokens": max_tokens,
                    "temperature": temperature,
                }

                if system_prompt is not None:
                    payload["system"] = system_prompt

                response = await client.post(
                    "/messages",
                    json=payload,
                    timeout=timeout,
                )
                response.raise_for_status()
                data = response.json()

                return data["content"][0]["text"]

            except httpx.HTTPError as e:
                if attempt == max_retries - 1:
                    raise RuntimeError(
                        f"Anthropic generation failed after {max_retries} retries: {e}"
                    ) from e

                delay = base_delay * (2**attempt)
                await asyncio.sleep(delay)

        return ""

    async def close(self) -> None:
        """Close the HTTP client and cleanup resources."""
        if self.client is not None:
            await self.client.aclose()
            self.client = None


class OpenRouterProvider:
    """LLM provider implementation using OpenRouter API.

    OpenRouter provides access to multiple LLM providers through an OpenAI-compatible API.
    Requires HTTP-Referer header for tracking.
    """

    def __init__(self, api_key: str, model: str = "openai/gpt-4o-mini"):
        """Initialize OpenRouter provider.

        Args:
            api_key: OpenRouter API key
            model: Model identifier in provider/model format (e.g., openai/gpt-4o-mini)
        """
        self.api_key = api_key
        self.model = model
        self.client: httpx.AsyncClient | None = None
        self.base_url = "https://openrouter.ai/api/v1"

    async def _get_client(self) -> httpx.AsyncClient:
        """Lazy initialize HTTP client with auth and referer headers."""
        if self.client is None:
            self.client = httpx.AsyncClient(
                base_url=self.base_url,
                headers={
                    "Authorization": f"Bearer {self.api_key}",
                    "HTTP-Referer": "menos",
                    "Content-Type": "application/json",
                },
            )
        return self.client

    async def generate(
        self,
        prompt: str,
        *,
        system_prompt: str | None = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
        timeout: float = 60.0,
    ) -> str:
        """Generate text using OpenRouter Chat Completions API.

        Args:
            prompt: The user message/prompt
            system_prompt: Optional system message to guide generation
            max_tokens: Maximum tokens to generate
            temperature: Sampling temperature (0.0-2.0)
            timeout: Request timeout in seconds

        Returns:
            Generated text response

        Raises:
            RuntimeError: If generation fails after all retries
        """
        max_retries = 3
        base_delay = 1.0

        for attempt in range(max_retries):
            try:
                client = await self._get_client()

                messages: list[dict[str, str]] = []
                if system_prompt is not None:
                    messages.append({"role": "system", "content": system_prompt})
                messages.append({"role": "user", "content": prompt})

                payload: dict[str, Any] = {
                    "model": self.model,
                    "messages": messages,
                    "max_tokens": max_tokens,
                    "temperature": temperature,
                }

                response = await client.post(
                    "/chat/completions",
                    json=payload,
                    timeout=timeout,
                )
                response.raise_for_status()
                data = response.json()

                return data["choices"][0]["message"]["content"]

            except httpx.HTTPError as e:
                if attempt == max_retries - 1:
                    raise RuntimeError(
                        f"OpenRouter generation failed after {max_retries} retries: {e}"
                    ) from e

                delay = base_delay * (2**attempt)
                await asyncio.sleep(delay)

        return ""

    async def close(self) -> None:
        """Close the HTTP client and cleanup resources."""
        if self.client is not None:
            await self.client.aclose()
            self.client = None


class FallbackProvider:
    """LLM provider that tries multiple providers in order until one succeeds."""

    def __init__(self, providers: list[tuple[str, LLMProvider]]):
        """Initialize with an ordered list of (name, provider) tuples.

        Args:
            providers: Ordered list of (name, provider) to try in sequence
        """
        if not providers:
            raise ValueError("FallbackProvider requires at least one provider")
        self.providers = providers

    async def generate(
        self,
        prompt: str,
        *,
        system_prompt: str | None = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
        timeout: float = 60.0,
    ) -> str:
        """Try each provider in order until one succeeds.

        Args:
            prompt: The user message/prompt
            system_prompt: Optional system message to guide generation
            max_tokens: Maximum tokens to generate
            temperature: Sampling temperature (0.0-2.0)
            timeout: Request timeout in seconds

        Returns:
            Generated text from the first successful provider

        Raises:
            RuntimeError: If all providers fail
        """
        errors: list[tuple[str, Exception]] = []

        for name, provider in self.providers:
            logger.info("FallbackProvider: trying '%s'", name)
            try:
                result = await provider.generate(
                    prompt,
                    system_prompt=system_prompt,
                    max_tokens=max_tokens,
                    temperature=temperature,
                    timeout=timeout,
                )
                if not result or not result.strip():
                    logger.warning("FallbackProvider: '%s' returned empty response", name)
                    errors.append((name, RuntimeError("Empty response")))
                    continue
                return result
            except Exception as exc:
                logger.warning("FallbackProvider: '%s' failed: %s", name, exc)
                errors.append((name, exc))

        summary = "; ".join(f"{n}: {e}" for n, e in errors)
        raise RuntimeError(f"All providers failed: {summary}")

    async def close(self) -> None:
        """Close all underlying providers."""
        for _, provider in self.providers:
            await provider.close()


class NoOpLLMProvider:
    """No-operation LLM provider for testing or disabling LLM features.

    Returns empty strings without making any network calls.
    """

    def __init__(self, api_key: str = "", model: str = "noop"):
        """Initialize NoOp provider.

        Args:
            api_key: Unused, accepted for interface compatibility
            model: Unused, accepted for interface compatibility
        """
        self.api_key = api_key
        self.model = model

    async def generate(
        self,
        prompt: str,
        *,
        system_prompt: str | None = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
        timeout: float = 60.0,
    ) -> str:
        """Return empty string without generating anything.

        Args:
            prompt: Ignored
            system_prompt: Ignored
            max_tokens: Ignored
            temperature: Ignored
            timeout: Ignored

        Returns:
            Empty string
        """
        return ""

    async def close(self) -> None:
        """No resources to cleanup."""
        pass


# Type assertions to ensure protocol compliance
_openai: LLMProvider = OpenAIProvider(api_key="test")
_anthropic: LLMProvider = AnthropicProvider(api_key="test")
_openrouter: LLMProvider = OpenRouterProvider(api_key="test")
_noop: LLMProvider = NoOpLLMProvider()
_fallback: LLMProvider = FallbackProvider([("noop", NoOpLLMProvider())])

File: services/core/src/vault/llm-providers.ts
import type { LlmProviderType, VaultConfig } from "./config";

export type LlmGenerationOptions = {
  systemPrompt?: string;
  maxTokens?: number;
  temperature?: number;
  timeout?: number;
  signal?: AbortSignal;
};

export type LlmGeneration = {
  text: string;
  inputTokens?: number;
  outputTokens?: number;
};

export type LlmProvider = {
  readonly model: string;
  generate(prompt: string, options?: LlmGenerationOptions): Promise<string>;
  close(): Promise<void>;
};

export type UsageReportingLlmProvider = LlmProvider & {
  generateWithUsage(prompt: string, options?: LlmGenerationOptions): Promise<LlmGeneration>;
};

export type LlmFetcher = (url: string, init?: RequestInit) => Promise<Response>;
export type LlmProviderPurpose = "expansion" | "synthesis" | "unifiedPipeline";

export type LlmFailureClassification = "transient" | "permanent";

export class LlmProviderFailure extends Error {
  readonly classification: LlmFailureClassification;
  readonly attempts: number;
  readonly status?: number;
  readonly lastError: unknown;

  constructor(message: string, details: { classification: LlmFailureClassification; attempts: number; status?: number; lastError: unknown }) {
    super(message, { cause: details.lastError });
    this.name = "LlmProviderFailure";
    this.classification = details.classification;
    this.attempts = details.attempts;
    this.status = details.status;
    this.lastError = details.lastError;
  }
}

export type LlmRetryOptions = {
  delay?: (milliseconds: number) => Promise<void>;
  random?: () => number;
  maxDelayMs?: number;
};

type JsonRecord = Record<string, unknown>;
type ProviderRequest = {
  path: string;
  payload: JsonRecord;
  headers: Record<string, string>;
};

type NormalizedLlmGenerationOptions = {
  systemPrompt: string | undefined;
  maxTokens: number;
  temperature: number;
  timeout: number;
  signal: AbortSignal | undefined;
};

const DEFAULT_MAX_TOKENS = 4096;
const DEFAULT_TEMPERATURE = 0.7;
const DEFAULT_TIMEOUT_SECONDS = 60;
const MAX_ATTEMPTS = 3;
const DEFAULT_MAX_BACKOFF_MS = 30_000;

function defaultFetcher(url: string, init?: RequestInit): Promise<Response> {
  return fetch(url, init);
}

function delay(milliseconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

function asRecord(value: unknown, provider: string): JsonRecord {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${provider} returned an invalid response`);
  }
  return value as JsonRecord;
}

function asArray(value: unknown, provider: string): unknown[] {
  if (!Array.isArray(value)) throw new Error(`${provider} returned an invalid response`);
  return value;
}

function requiredText(value: unknown, provider: string): string {
  if (typeof value !== "string") throw new Error(`${provider} returned an invalid response`);
  return value;
}

function tokenCount(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}

function parseRetryAfter(value: string): number | undefined {
  const seconds = Number(value);
  if (Number.isFinite(seconds) && seconds >= 0) return seconds * 1000;
  const timestamp = Date.parse(value);
  return Number.isNaN(timestamp) ? undefined : Math.max(0, timestamp - Date.now());
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

function generationOptions(options: LlmGenerationOptions): NormalizedLlmGenerationOptions {
  return {
    systemPrompt: options.systemPrompt ?? undefined,
    maxTokens: options.maxTokens ?? DEFAULT_MAX_TOKENS,
    temperature: options.temperature ?? DEFAULT_TEMPERATURE,
    timeout: options.timeout ?? DEFAULT_TIMEOUT_SECONDS,
    signal: options.signal,
  };
}

abstract class FetchLlmProvider implements UsageReportingLlmProvider {
  abstract readonly model: string;
  private readonly baseUrl: string;
  private readonly fetcher: LlmFetcher;
  private readonly retryOptions: Required<LlmRetryOptions>;

  protected constructor(baseUrl: string, fetcher: LlmFetcher, retryOptions: LlmRetryOptions = {}) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.fetcher = fetcher;
    this.retryOptions = {
      delay: retryOptions.delay ?? delay,
      random: retryOptions.random ?? Math.random,
      maxDelayMs: retryOptions.maxDelayMs ?? DEFAULT_MAX_BACKOFF_MS,
    };
  }

  protected abstract request(prompt: string, options: NormalizedLlmGenerationOptions): ProviderRequest;
  protected abstract parse(data: unknown): LlmGeneration;
  protected abstract readonly failurePrefix: string;

  async generate(prompt: string, options: LlmGenerationOptions = {}): Promise<string> {
    return (await this.generateWithUsage(prompt, options)).text;
  }

  async generateWithUsage(prompt: string, options: LlmGenerationOptions = {}): Promise<LlmGeneration> {
    const request = this.request(prompt, generationOptions(options));
    const normalized = generationOptions(options);
    const response = await this.post(request, normalized.timeout, normalized.signal);
    try {
      return this.parse(await response.json());
    } catch (error: unknown) {
      throw new LlmProviderFailure(`${this.failurePrefix} returned an invalid response`, { classification: "permanent", attempts: 1, lastError: error });
    }
  }

  async close(): Promise<void> {
    return Promise.resolve();
  }

  private async post(request: ProviderRequest, timeoutSeconds: number, callerSignal?: AbortSignal): Promise<Response> {
    let lastError: unknown;
    let lastStatus: number | undefined;
    for (let attempt = 1; attempt <= MAX_ATTEMPTS; attempt += 1) {
      if (callerSignal?.aborted) throw callerSignal.reason;
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), timeoutSeconds * 1000);
      const abortCaller = () => controller.abort(callerSignal?.reason);
      callerSignal?.addEventListener("abort", abortCaller, { once: true });
      try {
        const response = await this.fetcher(`${this.baseUrl}${request.path}`, { method: "POST", headers: request.headers, body: JSON.stringify(request.payload), signal: controller.signal });
        if (response.ok) return response;
        lastStatus = response.status;
        if (![408, 425, 429, 500, 502, 503, 504].includes(response.status)) {
          throw new LlmProviderFailure(`${this.failurePrefix} failed: HTTP ${response.status}`, { classification: "permanent", attempts: attempt, status: response.status, lastError: new Error(`HTTP ${response.status}`) });
        }
        lastError = new Error(`HTTP ${response.status}`);
        if (attempt < MAX_ATTEMPTS) {
          const retryAfter = response.headers.get("retry-after");
          const retryMs = retryAfter && (response.status === 429 || response.status === 503) ? parseRetryAfter(retryAfter) : undefined;
          await this.waitBeforeRetry(attempt, retryMs, callerSignal);
        }
      } catch (error: unknown) {
        if (callerSignal?.aborted) throw callerSignal.reason;
        if (error instanceof LlmProviderFailure) throw error;
        lastError = error;
        if (attempt < MAX_ATTEMPTS) await this.waitBeforeRetry(attempt, undefined, callerSignal);
      } finally {
        clearTimeout(timeout);
        callerSignal?.removeEventListener("abort", abortCaller);
      }
    }
    throw new LlmProviderFailure(`${this.failurePrefix} exhausted after ${MAX_ATTEMPTS} attempts: ${errorMessage(lastError)}`, { classification: "transient", attempts: MAX_ATTEMPTS, status: lastStatus, lastError });
  }

  private async waitBeforeRetry(attempt: number, retryAfter: number | undefined, signal?: AbortSignal): Promise<void> {
    if (signal?.aborted) throw signal.reason;
    const cap = Math.min(this.retryOptions.maxDelayMs, 1000 * (2 ** (attempt - 1)));
    const requestedDelay = retryAfter ?? this.retryOptions.random() * cap;
    const milliseconds = Math.min(cap, Math.max(0, requestedDelay));
    const wait = this.retryOptions.delay(milliseconds);
    if (signal === undefined) {
      await wait;
      return;
    }
    await Promise.race([
      wait,
      new Promise<void>((_, reject) => {
        const abort = () => reject(signal.reason);
        signal.addEventListener("abort", abort, { once: true });
        wait.finally(() => signal.removeEventListener("abort", abort)).catch(() => undefined);
      }),
    ]);
  }
}

export class OllamaLLMProvider extends FetchLlmProvider {
  readonly model: string;
  protected readonly failurePrefix = "LLM generation";

  constructor(baseUrl: string, model: string, fetcher: LlmFetcher = defaultFetcher, retryOptions?: LlmRetryOptions) {
    super(baseUrl, fetcher, retryOptions);
    this.model = model;
  }

  protected request(prompt: string, options: NormalizedLlmGenerationOptions): ProviderRequest {
    const payload: JsonRecord = {
      model: this.model,
      prompt,
      stream: false,
      options: {
        num_predict: options.maxTokens,
        temperature: options.temperature,
      },
    };
    if (options.systemPrompt !== undefined) payload.system = options.systemPrompt;
    return { path: "/api/generate", payload, headers: { "content-type": "application/json" } };
  }

  protected parse(data: unknown): LlmGeneration {
    const response = asRecord(data, "Ollama");
    const text = requiredText(response.response, "Ollama");
    return {
      text,
      inputTokens: tokenCount(response.prompt_eval_count),
      outputTokens: tokenCount(response.eval_count),
    };
  }
}

abstract class OpenAiCompatibleProvider extends FetchLlmProvider {
  protected request(prompt: string, options: NormalizedLlmGenerationOptions): ProviderRequest {
    const messages: JsonRecord[] = [];
    if (options.systemPrompt !== undefined) messages.push({ role: "system", content: options.systemPrompt });
    messages.push({ role: "user", content: prompt });
    return {
      path: "/chat/completions",
      payload: {
        model: this.model,
        messages,
        max_tokens: options.maxTokens,
        temperature: options.temperature,
      },
      headers: this.headers(),
    };
  }

  protected parse(data: unknown): LlmGeneration {
    const response = asRecord(data, this.failurePrefix);
    const firstChoice = asRecord(asArray(response.choices, this.failurePrefix)[0], this.failurePrefix);
    const message = asRecord(firstChoice.message, this.failurePrefix);
    const usage = response.usage === undefined ? undefined : asRecord(response.usage, this.failurePrefix);
    return {
      text: requiredText(message.content, this.failurePrefix),
      inputTokens: usage === undefined ? undefined : tokenCount(usage.prompt_tokens),
      outputTokens: usage === undefined ? undefined : tokenCount(usage.completion_tokens),
    };
  }

  protected abstract headers(): Record<string, string>;
}

export class OpenAIProvider extends OpenAiCompatibleProvider {
  readonly model: string;
  readonly apiKey: string;
  protected readonly failurePrefix = "OpenAI generation";

  constructor(apiKey: string, model = "gpt-4o-mini", fetcher: LlmFetcher = defaultFetcher, retryOptions?: LlmRetryOptions) {
    super("https://api.openai.com/v1", fetcher, retryOptions);
    this.apiKey = apiKey;
    this.model = model;
  }

  protected headers(): Record<string, string> {
    return { Authorization: `Bearer ${this.apiKey}`, "Content-Type": "application/json" };
  }
}

export class AnthropicProvider extends FetchLlmProvider {
  readonly model: string;
  readonly apiKey: string;
  protected readonly failurePrefix = "Anthropic generation";

  constructor(apiKey: string, model = "claude-3-5-haiku-20241022", fetcher: LlmFetcher = defaultFetcher, retryOptions?: LlmRetryOptions) {
    super("https://api.anthropic.com/v1", fetcher, retryOptions);
    this.apiKey = apiKey;
    this.model = model;
  }

  protected request(prompt: string, options: NormalizedLlmGenerationOptions): ProviderRequest {
    const payload: JsonRecord = {
      model: this.model,
      messages: [{ role: "user", content: prompt }],
      max_tokens: options.maxTokens,
      temperature: options.temperature,
    };
    if (options.systemPrompt !== undefined) payload.system = options.systemPrompt;
    return {
      path: "/messages",
      payload,
      headers: {
        "x-api-key": this.apiKey,
        "anthropic-version": "2023-06-01",
        "Content-Type": "application/json",
      },
    };
  }

  protected parse(data: unknown): LlmGeneration {
    const response = asRecord(data, "Anthropic");
    const firstContent = asRecord(asArray(response.content, "Anthropic")[0], "Anthropic");
    const usage = response.usage === undefined ? undefined : asRecord(response.usage, "Anthropic");
    return {
      text: requiredText(firstContent.text, "Anthropic"),
      inputTokens: usage === undefined ? undefined : tokenCount(usage.input_tokens),
      outputTokens: usage === undefined ? undefined : tokenCount(usage.output_tokens),
    };
  }
}

export class OpenRouterProvider extends OpenAiCompatibleProvider {
  readonly model: string;
  readonly apiKey: string;
  protected readonly failurePrefix = "OpenRouter generation";

  constructor(apiKey: string, model = "openai/gpt-4o-mini", fetcher: LlmFetcher = defaultFetcher, retryOptions?: LlmRetryOptions) {
    super("https://openrouter.ai/api/v1", fetcher, retryOptions);
    this.apiKey = apiKey;
    this.model = model;
  }

  protected headers(): Record<string, string> {
    return {
      Authorization: `Bearer ${this.apiKey}`,
      "HTTP-Referer": "menos",
      "Content-Type": "application/json",
    };
  }
}

export class NoOpLLMProvider implements UsageReportingLlmProvider {
  readonly apiKey: string;
  readonly model: string;

  constructor(apiKey = "", model = "noop") {
    this.apiKey = apiKey;
    this.model = model;
  }

  async generate(_prompt: string, _options: LlmGenerationOptions = {}): Promise<string> {
    return "";
  }

  async generateWithUsage(_prompt: string, _options: LlmGenerationOptions = {}): Promise<LlmGeneration> {
    return { text: "", inputTokens: 0, outputTokens: 0 };
  }

  async close(): Promise<void> {
    return Promise.resolve();
  }
}

export class FallbackProvider implements UsageReportingLlmProvider {
  readonly providers: readonly (readonly [string, UsageReportingLlmProvider])[];
  readonly model: string;

  constructor(providers: readonly (readonly [string, UsageReportingLlmProvider])[]) {
    if (providers.length === 0) throw new Error("FallbackProvider requires at least one provider");
    this.providers = providers;
    this.model = providers[0]?.[1].model ?? "unknown";
  }

  async generate(prompt: string, options: LlmGenerationOptions = {}): Promise<string> {
    return (await this.generateWithUsage(prompt, options)).text;
  }

  async generateWithUsage(prompt: string, options: LlmGenerationOptions = {}): Promise<LlmGeneration> {
    const errors: string[] = [];
    for (const [name, provider] of this.providers) {
      if (options.signal?.aborted) throw options.signal.reason;
      try {
        const result = await provider.generateWithUsage(prompt, options);
        if (result.text.trim() === "") {
          errors.push(`${name}: Error: Empty response`);
          continue;
        }
        return result;
      } catch (error: unknown) {
        if (options.signal?.aborted) throw options.signal.reason;
        errors.push(`${name}: ${errorMessage(error)}`);
      }
    }
    throw new Error(`All providers failed: ${errors.join("; ")}`);
  }

  async close(): Promise<void> {
    await Promise.all(this.providers.map(([, provider]) => provider.close()));
  }
}

export function providerName(provider: LlmProvider): string {
  if (provider instanceof OpenAIProvider) return "openai";
  if (provider instanceof AnthropicProvider) return "anthropic";
  if (provider instanceof OpenRouterProvider) return "openrouter";
  if (provider instanceof OllamaLLMProvider) return "ollama";
  return provider.constructor.name.toLowerCase();
}

export function buildOpenRouterChain(apiKey: string, model = "", fetcher: LlmFetcher = defaultFetcher): UsageReportingLlmProvider {
  if (apiKey === "") throw new Error("openrouter_api_key must be set for openrouter provider");
  if (model !== "") return new OpenRouterProvider(apiKey, model, fetcher);
  return new FallbackProvider([
    ["aurora", new OpenRouterProvider(apiKey, "openrouter/aurora-alpha", fetcher)],
    ["gpt-oss", new OpenRouterProvider(apiKey, "openai/gpt-oss-120b:free", fetcher)],
    ["deepseek", new OpenRouterProvider(apiKey, "deepseek/deepseek-r1-0528:free", fetcher)],
    ["gemma3", new OpenRouterProvider(apiKey, "google/gemma-3-27b-it:free", fetcher)],
  ]);
}

export function createLlmProvider(config: VaultConfig, provider: LlmProviderType, model: string, fetcher: LlmFetcher = defaultFetcher): UsageReportingLlmProvider {
  if (provider === "ollama") return new OllamaLLMProvider(config.ollamaUrl, model, fetcher);
  if (provider === "openai") {
    if (config.openaiApiKey === undefined) throw new Error("openai_api_key must be set when using openai provider");
    return new OpenAIProvider(config.openaiApiKey, model, fetcher);
  }
  if (provider === "anthropic") {
    if (config.anthropicApiKey === undefined) throw new Error("anthropic_api_key must be set when using anthropic provider");
    return new AnthropicProvider(config.anthropicApiKey, model, fetcher);
  }
  if (provider === "openrouter") return buildOpenRouterChain(config.openrouterApiKey ?? "", model, fetcher);
  return new NoOpLLMProvider();
}

export function createConfiguredLlmProvider(config: VaultConfig, purpose: LlmProviderPurpose, fetcher: LlmFetcher = defaultFetcher): UsageReportingLlmProvider {
  if (purpose === "expansion") return createLlmProvider(config, config.agentExpansionProvider, config.agentExpansionModel, fetcher);
  if (purpose === "synthesis") return createLlmProvider(config, config.agentSynthesisProvider, config.agentSynthesisModel, fetcher);
  return createLlmProvider(config, config.unifiedPipelineProvider, config.unifiedPipelineModel, fetcher);
}
